### ISSUE TITLE
AQL: STARTS WITH / ENDS WITH / CONTAINS operators

### ISSUE THREAD

**Issue by Simran-B** [arangodb/arangodb#1796]
Skiplist indices can be utilized to make "startsWith" queries fast, but it requires some tricky:

```js
FOR doc IN coll // {name: "John Deaux"}, {name: "Johnathan Rice"}
    FILTER doc.name >= "John" AND doc.name < "Joho"
    RETURN doc
```
Also see http://stackoverflow.com/questions/35587746/on-multiple-index-usage-in-arangodb

Unlike with full-text indexes, the proposed operators should work on whole strings, not individual words:

```js
FOR doc IN coll
    FILTER doc.name STARTS WITH "Ric"
    RETURN doc // should not match "Johnathan Rice"!
```

Indices should be utilized as much as possible, which may require an additional index type or configuration to index strings with the characters reversed, to facilitate "endsWith" searches (also see #1795).

Instead of `STARTS WITH` and `ENDS WITH`, the operators could follow the style of CSS 3 selectors (which might be inspired by regex):

- `doc.name ^= "John"` starts with
- `doc.name $= "aux"` ends with
- `doc.name ~= "nathan"` contains

The operators should also work without index on individual strings and obviate the need for STARTSWITH() etc. as proposed in #1045.

**Comment by iamacup** [comment]
Im not sure if this is the right issue for this - however - i am running this, which is awful in terms of performance:

```
FOR question IN questions FILTER question.easyLookupName == @easyLookupName
					FOR v, e IN 1..1 OUTBOUND question._id with_options FILTER v.validated == true AND LIKE(v.displayValue, @companyName, true)
						SORT v.displayValue ASC
				LIMIT @queryStartNumber, @resultsPerQuery
				RETURN {id: v._id, text: v.displayValue}
```

which, at the time of writing, i belive the LIKE operator did not use an index. I would love to have a better like operator - `doc.name ~= "nathan"` would transform this query into something like the below, which not only ready better, but may be more performant?

```
FOR question IN questions FILTER question.easyLookupName == @easyLookupName
					FOR v, e IN 1..1 OUTBOUND question._id with_options FILTER v.validated == true AND v.displayValue ~= @companyName
						SORT v.displayValue ASC
				LIMIT @queryStartNumber, @resultsPerQuery
				RETURN {id: v._id, text: v.displayValue}
```

**Comment by Simran-B** [comment]
@iamacup: LIKE() is not an operator, but a function. It creates a regular expression internally and then tests against the given input. The regex engine is "isolated" if you will, it can only access the input string, the regex pattern and possibly some regex flags, but nothing like ArangoDB indexes.

I wouldn't say that it's awfully slow, but of course slower than an index lookup, because it runs against all given documents. If you can filter by other criteria, so that only a hand full of a documents remain, then it's perfectly fine to be used - a dozen calls to LIKE() shouldn't hurt. You shouldn't use it as sole filter criterion however, because it would do a full collection scan and evaluate the regex for every single document.

While skiplist indexes could be exploited to achieve fast *starts with* / *ends with* searches (wildcard on either end of the search pattern), it would require a completed different index type to enable fast *contains* searches, such as suffix tries / suffix arrays (very memory consuming, e.g. 4 times the size of the data they index). MySQL uses the *Turbo Boyer-Moore* algorithm as far as I know, which is essentially a full collection scan, but with a smart per-record evaluation, which skips a lot of unnecessary character-by-character comparisons. If exact matching isn't a requirement, one could also look into techniques that combine fuzzy string matching with compression to achieve good performance in combination with space efficiency (Lucene/Solr does this AFAIK).

**Comment by JonDum** [comment]
It sounds like `CONTAINS` would be more difficult, but in the meantime it would be a huge win to have `STARTS WITH` and `ENDS WITH`, even if it's semantic sugar for the skiplist hackery... because it kinda stings that I have to jump through those kind of hoops to get something most other db's already have out of the box.

**Comment by robert-zaremba** [comment]
:+1: 
This would be very useful, as this functions will use indexes I suppose.

**Comment by MrCreosote** [comment]
Am I right in thinking that https://www.arangodb.com/docs/stable/aql/functions-string.html#starts_with will not use an index? At least in a query with 2 equality filters and one STARTS_WITH filter, the optimizer doesn't choose the compound index I set up with the 2 equality fields first and finally the STARTS_WITH field, but rather a different compound index that's similar, but the 3rd field is used in an UPDATE clause (where an index isn't needed...).

**Comment by Simran-B** [comment]
Correct, only https://www.arangodb.com/docs/stable/aql/functions-arangosearch.html#starts_with can utilize an index (inverted index/View)

**Comment by MrCreosote** [comment]
Belated thanks for the answer... sorry about that.

To follow on, the ArangoSearch `LIKE()` function also says it uses an index [1], which implies that ArangoSearch has some sort of suffix tree or equivalent data structure implemented? Is that right?

Will `LIKE()` work with an inverted index?

Thanks again

[1] https://www.arangodb.com/docs/stable/aql/functions-arangosearch.html#like
